hpl-2.2/man/man3/HPL_pdmxswp.3
.TH HPL_pdmxswp 3 "February 24, 2016" "HPL 2.2" "HPL Library Functions"
.SH NAME
HPL_pdmxswp \- swaps and broacast the pivot row.
.SH SYNOPSIS
\fB\&#include "hpl.h"\fR
 
\fB\&void\fR
\fB\&HPL_pdmxswp(\fR
\fB\&HPL_T_panel *\fR
\fI\&PANEL\fR,
\fB\&const int\fR
\fI\&M\fR,
\fB\&const int\fR
\fI\&II\fR,
\fB\&const int\fR
\fI\&JJ\fR,
\fB\&double *\fR
\fI\&WORK\fR
\fB\&);\fR
.SH DESCRIPTION
\fB\&HPL_pdmxswp\fR
swaps  and  broadcasts  the  absolute value max row using
bi-directional exchange.  The buffer is partially set by HPL_dlocmax.
 
Bi-directional  exchange  is  used  to  perform  the  swap::broadcast
operations  at once  for one column in the panel.  This  results in a
lower number of slightly larger  messages than usual.  On P processes
and assuming bi-directional links,  the running time of this function
can be approximated by
 
   log_2( P ) * ( lat + ( 2 * N0 + 4 ) / bdwth )
 
where  lat and bdwth are the latency and bandwidth of the network for
double precision real elements.  Communication  only  occurs  in  one
process  column. Mono-directional links  will cause the communication
cost to double.
.SH ARGUMENTS
.TP 8
PANEL   (local input/output)    HPL_T_panel *
On entry,  PANEL  points to the data structure containing the
panel information.
.TP 8
M       (local input)           const int
On entry,  M specifies the local number of rows of the matrix
column on which this function operates.
.TP 8
II      (local input)           const int
On entry, II  specifies the row offset where the column to be
operated on starts with respect to the panel.
.TP 8
JJ      (local input)           const int
On entry, JJ  specifies the column offset where the column to
be operated on starts with respect to the panel.
.TP 8
WORK    (local workspace)       double *
On entry, WORK  is a workarray of size at least 2 * (4+2*N0).
It  is assumed that  HPL_dlocmax  was called  prior  to  this
routine to  initialize  the first four entries of this array.
On exit, the  N0  length max row is stored in WORK[4:4+N0-1];
Note that this is also the  JJth  row  (or column) of L1. The
remaining part is used as a temporary array.
.SH SEE ALSO
.BR HPL_dlocmax \ (3),
.BR HPL_dlocswpN \ (3),
.BR HPL_dlocswpT \ (3),
.BR HPL_pdpancrN \ (3),
.BR HPL_pdpancrT \ (3),
.BR HPL_pdpanllN \ (3),
.BR HPL_pdpanllT \ (3),
.BR HPL_pdpanrlN \ (3),
.BR HPL_pdpanrlT \ (3),
.BR HPL_pdrpancrN \ (3),
.BR HPL_pdrpancrT \ (3),
.BR HPL_pdrpanllN \ (3),
.BR HPL_pdrpanllT \ (3),
.BR HPL_pdrpanrlN \ (3),
.BR HPL_pdrpanrlT \ (3),
.BR HPL_pdfact \ (3).


hpl-2.2/HISTORY
==============================================================
 High Performance Computing Linpack Benchmark (HPL)
 HPL - 2.2 - February 24, 2016
==============================================================

 History

 - 09/09/00 Public release of Version 1.0

 - 09/27/00 A couple of mistakes in the  VSIPL  port have been
 corrected.  The tar file as well as the web site were updated
 on September 27th, 2000.  Note  that  these problems were not
 affecting the BLAS version of the software in any way.

 - 01/01/04 Version 1.0a
 The  MPI  process grid  numbering  scheme  is now an run-time
 option.
 The inlined assembly  timer  routine that caused the compila-
 tion to fail when using  gcc  version 3.3  and above has been
 removed from the package.
 Various building problems on the T3E have been fixed;  Thanks
 to Edward Anderson.

 - 15/12/04 Version 1.0b
 Weakness of the pseudo-random matrix generator found for pro-
 blem sizes being power of twos and larger  than 2^15;  Thanks
 to Gregory Bauer. This problem has not been fixed. It is thus
 currently recommended to  HPL  users willing to test matrices
 of size larger than 2^15 to not use power twos.

 When the matrix size is such that one needs  > 16 GB  per MPI
 rank,  the  intermediate  calculation  (mat.ld+1) * mat.nq in
 HPL_pdtest.c  ends up  overflowing  because  it is done using
 32-bit arithmetic.  This issue has been fixed by  typecasting
 to size_t; Thanks to John Baron.

 - 09/10/08 Version 2.0

 Piotr Luszczek changed to 64-bit RNG, modified files:
 -- [M] include/hpl_matgen.h
 -- [M] testing/matgen/HPL_ladd.c
 -- [M] testing/matgen/HPL_lmul.c
 -- [M] testing/matgen/HPL_rand.c
 -- [M] testing/ptest/HPL_pdinfo.c

 For a motivation for the change, see:
    Dongarra and Langou, ``The Problem with the Linpack
    Benchmark Matrix Generator'', LAWN 206, June 2008.

 -- [M] testing/ptest/HPL_pdtest.c  --

 Julien Langou changed the test for correctness from
      ||Ax-b||_oo / ( eps * ||A||_1  * N            )
      ||Ax-b||_oo / ( eps * ||A||_1  * ||x||_1      )
      ||Ax-b||_oo / ( eps * ||A||_oo * ||x||_oo * N )
 to the normwise backward error
      || r ||_oo / ( eps * ( || x ||_oo * || A ||_oo + || b ||_oo ) * N )
 See:
  Nicholas J. Higham, ``Accuracy and Stability of Numerical Algorithms'',
  Society for Industrial and Applied Mathematics, Philadelphia, PA, USA,
  Second Edition, pages = xxx+680, ISBN = 0-89871-521-0, 2002.

 Note that in our case || b ||_oo is almost for sure
 1/2, we compute it anyway.

 - 10/26/2012 Version 2.1

 Piotr Luszczek introduced exact time stamping for HPL_pdgesv():
 -- [M] dist/include/hpl_misc.h
 -- [M] dist/testing/ptest/HPL_pdtest.c

 Piotr Luszczek fixed out-of-bounds access in data spreading functions
 and exact time stamping for HPL_pdgesv():
 -- [M] dist/src/pgesv/HPL_spreadN.c
 -- [M] dist/src/pgesv/HPL_spreadT.c
 Thanks to Stephen Whalen from Cray.

 - 02/24/2016 Version 2.2

 Piotr Luszczek added continuous reporting of factorization progress
 submitted by Intel and make scripts that uses Intel software tools and
 libraries and their Apple's Mac OS X equivalents.


hpl-2.2/src/pfact/HPL_pdmxswp.c
/* 
 * -- High Performance Computing Linpack Benchmark (HPL)                
 *    HPL - 2.2 - February 24, 2016                          
 *    Antoine P. Petitet                                                
 *    University of Tennessee, Knoxville                                
 *    Innovative Computing Laboratory                                 
 *    (C) Copyright 2000-2008 All Rights Reserved                       
 *                                                                      
 * -- Copyright notice and Licensing terms:                             
 *                                                                      
 * Redistribution  and  use in  source and binary forms, with or without
 * modification, are  permitted provided  that the following  conditions
 * are met:                                                             
 *                                                                      
 * 1. Redistributions  of  source  code  must retain the above copyright
 * notice, this list of conditions and the following disclaimer.        
 *                                                                      
 * 2. Redistributions in binary form must reproduce  the above copyright
 * notice, this list of conditions,  and the following disclaimer in the
 * documentation and/or other materials provided with the distribution. 
 *                                                                      
 * 3. All  advertising  materials  mentioning  features  or  use of this
 * software must display the following acknowledgement:                 
 * This  product  includes  software  developed  at  the  University  of
 * Tennessee, Knoxville, Innovative Computing Laboratory.             
 *                                                                      
 * 4. The name of the  University,  the name of the  Laboratory,  or the
 * names  of  its  contributors  may  not  be used to endorse or promote
 * products  derived   from   this  software  without  specific  written
 * permission.                                                          
 *                                                                      
 * -- Disclaimer:                                                       
 *                                                                      
 * THIS  SOFTWARE  IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,  INCLUDING,  BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY
 * OR  CONTRIBUTORS  BE  LIABLE FOR ANY  DIRECT,  INDIRECT,  INCIDENTAL,
 * SPECIAL,  EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES  (INCLUDING,  BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT,  STRICT LIABILITY,  OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 * ---------------------------------------------------------------------
 */ 
/*
 * Include files
 */
#include "hpl.h"

#ifdef STDC_HEADERS
void HPL_pdmxswp
(
   HPL_T_panel *                    PANEL,
   const int                        M,
   const int                        II,
   const int                        JJ,
   double *                         WORK
)
#else
void HPL_pdmxswp
( PANEL, M, II, JJ, WORK )
   HPL_T_panel *                    PANEL;
   const int                        M;
   const int                        II;
   const int                        JJ;
   double *                         WORK;
#endif
{
/* 
 * Purpose
 * =======
 *
 * HPL_pdmxswp swaps  and  broadcasts  the  absolute value max row using
 * bi-directional exchange.  The buffer is partially set by HPL_dlocmax.
 *  
 * Bi-directional  exchange  is  used  to  perform  the  swap::broadcast
 * operations  at once  for one column in the panel.  This  results in a
 * lower number of slightly larger  messages than usual.  On P processes
 * and assuming bi-directional links,  the running time of this function
 * can be approximated by
 *  
 *    log_2( P ) * ( lat + ( 2 * N0 + 4 ) / bdwth )
 *  
 * where  lat and bdwth are the latency and bandwidth of the network for
 * double precision real elements.  Communication  only  occurs  in  one
 * process  column. Mono-directional links  will cause the communication
 * cost to double.
 *
 * Arguments
 * =========
 *
 * PANEL   (local input/output)          HPL_T_panel *
 *         On entry,  PANEL  points to the data structure containing the
 *         panel information.
 *
 * M       (local input)                 const int
 *         On entry,  M specifies the local number of rows of the matrix
 *         column on which this function operates.
 *
 * II      (local input)                 const int
 *         On entry, II  specifies the row offset where the column to be
 *         operated on starts with respect to the panel.
 *
 * JJ      (local input)                 const int
 *         On entry, JJ  specifies the column offset where the column to
 *         be operated on starts with respect to the panel.
 *
 * WORK    (local workspace)             double *
 *         On entry, WORK  is a workarray of size at least 2 * (4+2*N0).
 *         It  is assumed that  HPL_dlocmax  was called  prior  to  this
 *         routine to  initialize  the first four entries of this array.
 *         On exit, the  N0  length max row is stored in WORK[4:4+N0-1];
 *         Note that this is also the  JJth  row  (or column) of L1. The
 *         remaining part is used as a temporary array.
 *
 * ---------------------------------------------------------------------
 */ 
/*
 * .. Local Variables ..
 */
   double                     gmax, tmp1;
   double                     * A0, * Wmx, * Wwork;
   HPL_T_grid                 * grid;
   MPI_Comm                   comm;
   unsigned int               hdim, ip2, ip2_, ipow, k, mask;
   int                        Np2, cnt_, cnt0, i, icurrow, lda, mydist,
                              mydis_, myrow, n0, nprow, partner, rcnt,
                              root, scnt, size_;
/* ..
 * .. Executable Statements ..
 */
#ifdef HPL_DETAILED_TIMING
   HPL_ptimer( HPL_TIMING_MXSWP );
#endif
   grid = PANEL->grid; myrow = grid->myrow; nprow = grid->nprow;
/*
 * ip2  : the smallest power of two less than or equal to nprow;
 * hdim : dimension of the hypercube made of those ip2 processes;
 * Np2  : logical flag indicating whether or not nprow is a power of 2;
 */
   comm    = grid->col_comm; ip2 = (unsigned int)(grid->row_ip2);
   hdim    = (unsigned int)(grid->row_hdim);     n0  = PANEL->jb;
   icurrow = PANEL->prow; Np2 = (int)( ( size_ = nprow - ip2 ) != 0 );
   mydist  = MModSub( myrow, icurrow, nprow );
/*
 * Set up pointers in workspace:  WORK and Wwork  point to the beginning
 * of the buffers of size 4 + 2*N0 to be combined. Wmx points to the row
 * owning the local (before combine) and global (after combine) absolute
 * value max. A0 points to the copy of the current row of the matrix.
 */
   cnt0  = ( cnt_ = n0 + 4 ) + n0; A0 = ( Wmx = WORK + 4 ) + n0;
   Wwork = WORK + cnt0;
/*
 * Wmx[0:N0-1] := A[ilindx,0:N0-1] where ilindx is  (int)(WORK[1])  (row
 * with max in current column). If I am the current process row, pack in
 * addition the current row of A in A0[0:N0-1].  If I do not own any row
 * of A, then zero out Wmx[0:N0-1].
 */
   if( M > 0 )
   {
      lda = PANEL->lda;
      HPL_dcopy( n0, Mptr( PANEL->A, II+(int)(WORK[1]), 0, lda ), lda,
                 Wmx, 1 );
      if( myrow == icurrow )
      { HPL_dcopy( n0, Mptr( PANEL->A, II, 0, lda ), lda, A0, 1 ); }
   }
   else { for( i = 0; i < n0; i++ ) Wmx[i] = HPL_rzero; }
/*
 * Combine the results (bi-directional exchange):  the process coordina-
 * tes are relative to icurrow,  this allows to reduce the communication
 * volume when nprow is not a power of 2.
 *
 * When nprow is not a power of 2:  proc[i-ip2] receives local data from
 * proc[i]  for all i in [ip2..nprow).  In addition,  proc[0]  (icurrow)
 * sends to proc[ip2] the current row of A  for later broadcast in procs
 * [ip2..nprow).
 */
   if( ( Np2 != 0 ) &&
       ( ( partner = (int)((unsigned int)(mydist) ^ ip2 ) ) < nprow ) )
   {
      if( ( mydist & ip2 ) != 0 )
      {
         if( mydist == (int)(ip2) )
            (void) HPL_sdrv( WORK, cnt_, MSGID_BEGIN_PFACT, A0, n0,
                             MSGID_BEGIN_PFACT, MModAdd( partner,
                             icurrow, nprow ), comm );
         else
            (void) HPL_send( WORK, cnt_, MModAdd( partner, icurrow,
                             nprow ), MSGID_BEGIN_PFACT, comm );
      }
      else
      {
         if( mydist == 0 )
            (void) HPL_sdrv( A0, n0, MSGID_BEGIN_PFACT, Wwork, cnt_,
                             MSGID_BEGIN_PFACT, MModAdd( partner,
                             icurrow, nprow ), comm );
         else
            (void) HPL_recv( Wwork, cnt_, MModAdd( partner, icurrow,
                             nprow ), MSGID_BEGIN_PFACT, comm );
 
         tmp1 = Mabs( Wwork[0] ); gmax = Mabs( WORK[0] );
         if( ( tmp1 > gmax ) ||
             ( ( tmp1 == gmax ) && ( Wwork[3] < WORK[3] ) ) )
         { HPL_dcopy( cnt_, Wwork, 1, WORK, 1 ); }
      }
   }

   if( mydist < (int)(ip2) )
   {
/*
 * power of 2 part of the processes collection: processes  [0..ip2)  are
 * combining (binary exchange); proc[0] has two rows to send, but one to
 * receive.  At every step  k  in [0..hdim) of the algorithm,  a process 
 * pair exchanging 2 rows is such that  myrow >> k+1 is 0.  Among  those
 * processes the ones  that are sending one more row than  what they are
 * receiving are such that myrow >> k is equal to 0.
 */
      k = 0; ipow = 1;
 
      while( k < hdim )
      {
         if( ( (unsigned int)(mydist) >> ( k + 1 ) ) == 0 )
         {
            if( ( (unsigned int)(mydist) >> k ) == 0 )
            { scnt = cnt0; rcnt = cnt_; }
            else
            { scnt = cnt_; rcnt = cnt0; }
         }
         else { scnt = rcnt = cnt_; }
 
         partner = (int)( (unsigned int)(mydist) ^ ipow );
         (void) HPL_sdrv( WORK, scnt, MSGID_BEGIN_PFACT, Wwork, rcnt,
                          MSGID_BEGIN_PFACT, MModAdd( partner, icurrow,
                          nprow ), comm );
 
         tmp1 = Mabs( Wwork[0] ); gmax = Mabs( WORK[0] );
         if( ( tmp1 > gmax ) ||
             ( ( tmp1 == gmax ) && ( Wwork[3] < WORK[3] ) ) )
         {
            HPL_dcopy( ( rcnt == cnt0 ? cnt0 : cnt_ ), Wwork, 1,
                       WORK, 1 );
         }
         else if( rcnt == cnt0 )
         { HPL_dcopy( n0, Wwork+cnt_, 1, A0, 1 ); }
 
         ipow <<= 1; k++;
      }
   }
   else if( size_ > 1 )
   {
/*
 * proc[ip2] broadcast current row of A to procs [ip2+1..nprow).
 */
      k = (unsigned int)(size_) - 1; ip2_ = mask = 1;
      while( k > 1 ) { k >>= 1; ip2_ <<= 1; mask <<= 1; mask++; }
 
      root   = MModAdd( icurrow, (int)(ip2), nprow );
      mydis_ = MModSub( myrow,   root,       nprow );
 
      do
      {
         mask ^= ip2_;
         if( ( mydis_ & mask ) == 0 )
         {
            partner = (int)(mydis_ ^ ip2_);
            if( ( mydis_ & ip2_ ) != 0 )
            {
               (void) HPL_recv( A0, n0, MModAdd( root, partner,
                                nprow ), MSGID_BEGIN_PFACT, comm );
            }
            else if( partner < size_ )
            {
               (void) HPL_send( A0, n0, MModAdd( root, partner,
                                nprow ), MSGID_BEGIN_PFACT, comm );
            }
         }
         ip2_ >>= 1;
      } while( ip2_ > 0 );
   }
/*
 * If nprow is not a power of 2,  for all i in [ip2..nprow), proc[i-ip2]
 * sends the pivot row to proc[i]  along  with the first four entries of
 * the WORK array.
 */
   if( ( Np2 != 0 ) &&
       ( ( partner = (int)((unsigned int)(mydist) ^ ip2 ) ) < nprow ) )
   {
      if( ( mydist & ip2 ) != 0 )
      {
         (void) HPL_recv( WORK, cnt_, MModAdd( partner, icurrow,
                          nprow ), MSGID_BEGIN_PFACT, comm );
      }
      else
      {
         (void) HPL_send( WORK, cnt_, MModAdd( partner, icurrow,
                          nprow ), MSGID_BEGIN_PFACT, comm );
      }
   }
/*
 * Save the global pivot index in pivot array
 */
   (PANEL->DPIV)[JJ] = WORK[2];
#ifdef HPL_DETAILED_TIMING
   HPL_ptimer( HPL_TIMING_MXSWP );
#endif
/*
 * End of HPL_pdmxswp
 */
}
